File: auth0/src/main/java/com/auth0/android/provider/LogoutManager.java
package com.auth0.android.provider;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.util.Log;

import com.auth0.android.Auth0;
import com.auth0.android.Auth0Exception;

import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("WeakerAccess")
class LogoutManager extends ResumableManager {

    private static final String TAG = LogoutManager.class.getSimpleName();

    private static final String KEY_CLIENT_ID = "client_id";
    private static final String KEY_TELEMETRY = "auth0Client";
    private static final String KEY_RETURN_TO_URL = "returnTo";

    private final Auth0 account;
    private final VoidCallback callback;
    private final Map<String, String> parameters;

    private CustomTabsOptions ctOptions;

    LogoutManager(@NonNull Auth0 account, @NonNull VoidCallback callback, @NonNull String returnToUrl, @NonNull CustomTabsOptions ctOptions) {
        this.account = account;
        this.callback = callback;
        this.parameters = new HashMap<>();
        this.parameters.put(KEY_RETURN_TO_URL, returnToUrl);
        this.ctOptions = ctOptions;
    }

    void setCustomTabsOptions(@NonNull CustomTabsOptions options) {
        this.ctOptions = options;
    }

    void startLogout(Context context) {
        addClientParameters(parameters);
        Uri uri = buildLogoutUri();

        AuthenticationActivity.authenticateUsingBrowser(context, uri, ctOptions);
    }

    @Override
    boolean resume(AuthorizeResult result) {
        if (result.isCanceled()) {
            Auth0Exception exception = new Auth0Exception("The user closed the browser app so the logout was cancelled.");
            callback.onFailure(exception);
        } else {
            callback.onSuccess(null);
        }
        return true;
    }

    private Uri buildLogoutUri() {
        Uri logoutUri = Uri.parse(account.getLogoutUrl());
        Uri.Builder builder = logoutUri.buildUpon();
        for (Map.Entry<String, String> entry : parameters.entrySet()) {
            builder.appendQueryParameter(entry.getKey(), entry.getValue());
        }
        Uri uri = builder.build();
        logDebug("Using the following Logout URI: " + uri.toString());
        return uri;
    }

    private void addClientParameters(Map<String, String> parameters) {
        if (account.getTelemetry() != null) {
            parameters.put(KEY_TELEMETRY, account.getTelemetry().getValue());
        }
        parameters.put(KEY_CLIENT_ID, account.getClientId());
    }

    @VisibleForTesting
    CustomTabsOptions customTabsOptions() {
        return ctOptions;
    }

    private void logDebug(String message) {
        if (account.isLoggingEnabled()) {
            Log.d(TAG, message);
        }
    }
}

File: auth0/src/main/java/com/auth0/android/provider/LogoutManager.kt
package com.auth0.android.provider

import android.content.Context
import android.net.Uri
import android.util.Log
import com.auth0.android.Auth0
import com.auth0.android.authentication.AuthenticationException
import com.auth0.android.callback.Callback
import java.lang.ref.WeakReference
import java.util.*

internal class LogoutManager(
    private val account: Auth0,
    callback: Callback<Void?, AuthenticationException>,
    returnToUrl: String,
    ctOptions: CustomTabsOptions,
    federated: Boolean = false,
    private val launchAsTwa: Boolean = false,
    private val customLogoutUrl: String? = null
) : ResumableManager() {
    private val callbackRef = WeakReference(callback)

    private fun deliverSuccess() {
        val cb = callbackRef.get()
        if (cb != null) {
            cb.onSuccess(null)
        } else {
            WebAuthProvider.pendingLogoutResult =
                WebAuthProvider.PendingResult.Success(null)
        }
    }

    private fun deliverFailure(error: AuthenticationException) {
        val cb = callbackRef.get()
        if (cb != null) {
            cb.onFailure(error)
        } else {
            WebAuthProvider.pendingLogoutResult =
                WebAuthProvider.PendingResult.Failure(error)
        }
    }
    private val parameters: MutableMap<String, String>
    private val ctOptions: CustomTabsOptions
    fun startLogout(context: Context) {
        addClientParameters(parameters)
        val uri = buildLogoutUri()
        AuthenticationActivity.authenticateUsingBrowser(context, uri, launchAsTwa, ctOptions)
    }

    public override fun resume(result: AuthorizeResult): Boolean {
        if (result.isCanceled) {
            val exception = AuthenticationException(
                AuthenticationException.ERROR_VALUE_AUTHENTICATION_CANCELED,
                "The user closed the browser app so the logout was cancelled."
            )
            deliverFailure(exception)
        } else {
            deliverSuccess()
        }
        return true
    }

    override fun failure(exception: AuthenticationException) {
        deliverFailure(exception)
    }

    private fun buildLogoutUri(): Uri {
        val urlToUse = customLogoutUrl ?: account.logoutUrl
        val logoutUri = Uri.parse(urlToUse)
        val builder = logoutUri.buildUpon()
        for ((key, value) in parameters) {
            builder.appendQueryParameter(key, value)
        }
        val uri = builder.build()
        Log.d(TAG, "Using the following Logout URI: $uri")
        return uri
    }

    private fun addClientParameters(parameters: MutableMap<String, String>) {
        parameters[KEY_USER_AGENT] = account.auth0UserAgent.value
        parameters[KEY_CLIENT_ID] = account.clientId
    }

    companion object {
        private val TAG = LogoutManager::class.java.simpleName
        private const val KEY_CLIENT_ID = "client_id"
        private const val KEY_USER_AGENT = "auth0Client"
        private const val KEY_RETURN_TO_URL = "returnTo"
        private const val KEY_FEDERATED = "federated"
    }

    init {
        parameters = HashMap()
        parameters[KEY_RETURN_TO_URL] = returnToUrl
        if (federated) {
            // null or empty values are not included in the request
            parameters[KEY_FEDERATED] = "1"
        }
        this.ctOptions = ctOptions
    }
}